### PR TITLE
Fix database view for data out

### LINKED ISSUE #34: Database view adjustments for ega-data-api
### Description
Current database view that is meant for use by EGA-data-API cannot be used as is and requires changes some changes in retrieving the data by the stable id.
e.g. in https://github.com/EGA-archive/LocalEGA/blob/master/deploy/images/db/ebi.sql#L5
the `file_id` is integer and corresponds to an internal ID used by LocalEGA to store file
while in https://github.com/EGA-archive/ega-data-api/blob/master/ega-data-api-filedatabase/src/main/resources/File.sql#L22 `file_id` was designed to be used as the `stable_id`

There are also several columns missing from https://github.com/EGA-archive/LocalEGA/blob/master/deploy/images/db/ebi.sql such as `checksum`, `checksum_type` `display_file_name`, `file_status` (even though it seems that the view is materialised with `READY` files) `file_type` does not seem to be used by `ega-data-api`

Current solution developed in `test/m4` branch of EGA-Data-API, changes the `file_id` in ega-data-api to Integer and only allows retrieval of a file with the `file_id` and not the `stable_id`.

Should files be retrievable from RES with `file_id` or `stable_id`?
RES connects directly to archive/vault.

In principle we should also consider the notes here: https://github.com/EGA-archive/ega-data-api/issues/10#issuecomment-407743725 when designing the view.

### Definition of Done
Design a view that will facilitate retrieval of archived files via their `stable_id` without adding too much overhead. Files can be retrieved with `stable_id` from DataEdge.

### How to test
Peer review and the view can be used as is by ega-data-api without huge changes to the source code.

### PR DESCRIPTION
**Pull request by omllobet**
### Describe the pull request:

- [X] Bug fix
- [ ] Functional change
- [ ] New feature
- [X] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

This commit changes the views from LocalEga to be compatible with
the current data-out code

### Related issues:

This PR fixes #34 

### Additional information:

I cannot include any tests right now until other issues are fixed. Also there is a new issue for the data out code because to access the s3 backend it uses the fileid instead of the filepath (https://github.com/EGA-archive/ega-data-api/issues/101)

### CODE REVIEW

File: deploy/images/db/grants.sql [hunk 1]
```diff
@@ -13,15 +13,10 @@ CREATE USER lega_in;
 CREATE USER lega_out;
 
 -- Set up rights access for local_ega schema
-GRANT USAGE ON SCHEMA local_ega TO lega_in, lega_out;
+GRANT USAGE ON SCHEMA local_ega TO lega_in;
```
Comment (silverdaz) [inline deploy/images/db/grants.sql]: Are you sure lega_out does not need usage on local_ega schema in order to see the main table?

File: deploy/images/db/ebi.sql [hunk 2]
```diff
@@ -1,85 +1,73 @@
...
+       status                                   AS file_status,
+       header                                   AS header
 FROM local_ega.main
 WHERE status = 'READY';
 
-
--- Relation File <-> Index File
-CREATE TABLE local_ega.index_files (
+-- Relation File EGAF <-> Dataset EGAD
+CREATE TABLE local_ega.filedataset (
```
Comment (silverdaz) [inline deploy/images/db/ebi.sql]: Not found of having the filedataset table in the "main" schema, as I don't think that table should even be existing.

Why the need to the table here, and a view a little below in the local_ega_download schema?

File: deploy/images/db/ebi.sql [hunk 3]
```diff
@@ -1,85 +1,73 @@
...
 
-       file_id  INTEGER NOT NULL REFERENCES local_ega.main (id) ON DELETE CASCADE,
+-- This view was created to be in sync with the entity eu.elixir.ega.ebi.downloader.domain.entity.FileDataset
+-- which uses a view and has an @Id annotation in file_id
+CREATE VIEW local_ega_download.file_dataset AS
+SELECT m.stable_id AS file_id, dataset_stable_id as dataset_id FROM local_ega.filedataset fd
+INNER JOIN local_ega.main m ON fd.file_id=m.id;
 
+-- Relation File <-> Index File
+CREATE TABLE local_ega.fileindexfile (
```
Comment (silverdaz) [inline deploy/images/db/ebi.sql]: Same as above

File: deploy/images/db/ebi.sql [hunk 4]
```diff
@@ -1,85 +1,73 @@
...
-);
+-- This view was created to be in sync with the entity eu.elixir.ega.ebi.downloader.domain.entity.FileIndexFile
+-- which seems to use a view and has an @Id annotation in file_id
+CREATE VIEW local_ega_download.file_index_file AS
+SELECT m.stable_id AS file_id, index_file_id FROM local_ega.fileindexfile fif
+INNER JOIN local_ega.main m ON fif.file_id=m.id;
 
 -- Event
-CREATE TABLE local_ega.event (
+CREATE TABLE local_ega_download.event (
```
Comment (silverdaz) [inline deploy/images/db/ebi.sql]: Instead of fixing the java code, to work with the newly organized table, we are dumping the data in the old tables?

Is that a quick fix to get things running or _the_ chosen solution?
In the latter case, I think it's not the way to go.

File: deploy/images/db/ebi.sql [hunk 5]
```diff
@@ -1,85 +1,73 @@
...
-
        created timestamp NOT NULL DEFAULT now(),
-
        CONSTRAINT event_pkey PRIMARY KEY (event_id)
-
 );
 
 -- Download Log
-CREATE TABLE local_ega.download_log (
+CREATE TABLE local_ega_download.download_log (
```
Comment (silverdaz) [inline deploy/images/db/ebi.sql]: Same as above

**Comment by silverdaz** [comment]
Reminder: This is the solution we implemented some time ago: 
https://github.com/EGA-archive/LocalEGA/blob/dev/deploy/images/db/download.sql

File: deploy/images/db/ebi.sql (hunk 2 above)
Comment (silverdaz) [reply deploy/images/db/ebi.sql]: Ideally it could even be a different database, so I don't see the need to mix it here.

Comment (silverdaz) [reply deploy/images/db/ebi.sql]: Hold on a little bit longer, until Oscar brings the modifications we talked about yesterday.
I think it'll make the database structure clearer, and overall better.

Comment (omllobet) [reply deploy/images/db/ebi.sql]: Moved that to it's own schema. The table has a reference to the id number while the data-out view needs the stable_id.

File: deploy/images/db/ebi.sql [hunk 6]
```diff
@@ -1,85 +1,47 @@
...
-       index_file_reference      TEXT NOT NULL,     -- file path if POSIX, object id if S3
-       index_file_type           local_ega.storage  -- S3 or POSIX file system
-);
-
 -- Relation File EGAF <-> Dataset EGAD
-CREATE TABLE local_ega.file2dataset (
+CREATE TABLE local_ega_ebi.filedataset (
        id       SERIAL, PRIMARY KEY(id), UNIQUE (id),
-
        file_id     INTEGER NOT NULL REFERENCES local_ega.main (id) ON DELETE CASCADE, -- not stable_id
```
Comment (blankdots) [inline deploy/images/db/ebi.sql]: I think here should be the `stable_id`, because when there is a request for a file via DataEdge for a file it is done via the `stable_id` e.g. 
```
curl -G -H "Authorization: Bearer $TOKEN" https://localhost:9090/files/EGA798762706
```

DataEdge makes a request in `FileInfoServiceImpl.java` to `getFilePermissionsEntity(String stableId)` and uses that stableID to call `FileDatabase/file/{stableId}/datasets` which in return looks into the view `local_ega_ebi.file_dataset` which gets data from this table - hopefully I got this right.

Long story short - we will be exposing internal file ids and we won't be able to retrieve a file with its stable_id from DataEdge if we do it like this.

File: deploy/images/db/download.sql [hunk 7]
```diff
@@ -57,106 +39,84 @@ DECLARE
      rid  INTEGER;
 BEGIN
 
+     -- Find the file
      SELECT * INTO archive_rec FROM local_ega.archive_files WHERE stable_id = sid LIMIT 1;
 
      IF archive_rec IS NULL THEN
-     	RAISE EXCEPTION 'Archive file not found for stable_id: % ', sid;
+     	RAISE EXCEPTION 'archive file not found for stable_id: % ', sid;
```
Comment (blankdots) [inline deploy/images/db/download.sql]: maybe this?
```suggestion
     	RAISE EXCEPTION 'archived file not found for stable_id: % ', sid;
```

File: deploy/images/db/ebi.sql (hunk 6 above)
Comment (blankdots) [reply deploy/images/db/ebi.sql]: I assume some additional logic might need implementing in `ega-data-api` to verify that the file with that stableID is there in `ebi_files` ...

Comment (omllobet) [reply deploy/images/db/ebi.sql]: The download_log and event log are now in local_ega_download.requests local_ega_download.success and local_ega_download.errors.

So the table links with the file_id of the main table but the file_dataset view (not filedataset table) has the stable_id.

To test it you with ega-data-api with current master you have to do 2 things:
* modify the code in 
the lines: 
from
```
    new SqlOutParameter("vault_path", Types.VARCHAR),
    new SqlOutParameter("valut_type", Types.OTHER),
```
to
```
    new SqlOutParameter("archive_path", Types.VARCHAR),
    new SqlOutParameter("archive_type", Types.OTHER),
```

Start ega-data-api with these profiles:

dataedge: no-oss,logger-log,add-user-ip-headers
res: no-oss,LocalEGA,log-transfer,repo-logger,db-repo-logger

Also for res you have to configure the database connection:
```
spring.datasource.url= jdbc:postgresql://localhost:5432/lega
spring.datasource.username=lega_out
spring.datasource.password=db_password
spring.jpa.properties.hibernate.default_schema = local_ega_ebi

## DataSource common settings
# Keep the connection alive if idle for a long time (needed in production)
spring.datasource.hikari.test-while-idle = true
spring.datasource.hikari.validation-query = SELECT 1
spring.datasource.hikari.initial-size = 3
spring.datasource.hikari.maximum-pool-size = 8
spring.datasource.hikari.minimum-idle = 1
```
About "Also not 100% sure how the new structure will satisfy the need for audit mentioned here: EGA-archive/ega-data-api#10 (comment)" Not sure I see what you mean. We can talk on slack if you want.

File: deploy/images/db/download.sql (hunk 7 above)
Comment (omllobet) [reply deploy/images/db/download.sql]: Sure, will update the code and push it today

File: deploy/images/db/ebi.sql (hunk 6 above)
Comment (blankdots) [reply deploy/images/db/ebi.sql]: > So the table links with the file_id of the main table but the file_dataset view (not filedataset table) has the stable_id.

Yes you are right I missed that.
Will try to test ASAP.

**Comment by silverdaz** [comment]
Quick comment: Is there a reason why comments here mention the java code?
This PR is about the database, and Java code discussions should go in the java-data-out repo.
And in here, we should link to the other repo's PR.

**Comment by blankdots** [comment]
@silverdaz I mentioned it because the issue this PR addressed mentioned integration with JAVA code.
